### PR TITLE
feat: /proc/mdstat collection

### PR DESCRIPTION
**Pull request by johnseekins**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.

This adds a new input that collects stats on Linux MD arrays on a system. It is based off https://github.com/prometheus/procfs/blob/master/mdstat.go. (which is apache licensed vs. telegraf's MIT.

### CODE REVIEW

**Comment by johnseekins** [comment]
Not really sure how to fix this lint complaints:
```
Warning: function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
Warning: function-result-limit: maximum number of return results per function exceeded; max 3 but got 5 (revive)
```
There are other input plugins that _also_ violate this limit (e.g. `kernel_vmstat` has a function with 4 return statements) so I'm struggling to see how this check is valid.

**Comment by johnseekins** [comment]
Sorry about all the commits, but `make test` doesn't seem to work correctly for me locally, so I'm leaning on circle to see my tests pass.

**Comment by johnseekins** [comment]
Looks like #9061 also implements this. I guess either solution works, I'm inclined towards this one (obviously), but I hope we add one soon.

**Comment by johnseekins** [comment]
The golang linter _must_ be improperly configured:
```
$ /home/vagrant/go/bin/golangci-lint run --out-format=github-actions --new=false --new-from-rev= | grep function-result-limit
::warning file=internal/templating/engine.go,line=22,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=internal/templating/template.go,line=19,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/parsers/graphite/parser.go,line=182,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/inputs/intel_rdt/publisher.go,line=67,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/inputs/intel_rdt/publisher.go,line=112,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 5 (revive)
::warning file=plugins/inputs/intel_rdt/publisher.go,line=158,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 5 (revive)
::warning file=plugins/inputs/intel_rdt/intel_rdt.go,line=464,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/inputs/postfix/postfix.go,line=35,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/inputs/intel_powerstat/intel_powerstat_test.go,line=482,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/inputs/proxmox/proxmox.go,line=239,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/inputs/smart/smart.go,line=553,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/inputs/smart/smart.go,line=567,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 4 (revive)
::warning file=plugins/inputs/aliyuncms/discovery.go,line=295,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 5 (revive)
::warning file=plugins/inputs/snmp/snmp.go,line=754,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 5 (revive)
::warning file=plugins/inputs/snmp/snmp.go,line=771,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 5 (revive)
::warning file=plugins/inputs/snmp/snmp.go,line=842,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 5 (revive)
::warning file=plugins/inputs/snmp/snmp.go,line=868,col=1::function-result-limit: maximum number of return results per function exceeded; max 3 but got 5 (revive)
```

File: plugins/inputs/mdstat/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+For a better idea of what each field represents, see the 
+[mdstat man page](https://raid.wiki.kernel.org/index.php/Mdstat).
+
+Stat collection based on Prometheus' mdstat collection library at https://github.com/prometheus/procfs/blob/master/mdstat.go
+
+
+### Configuration:
+
+```toml
+# Get kernel statistics from /proc/vmstat
```
Comment (ssoroka) [inline plugins/inputs/mdstat/README.md:15]: vmstat? copy pasted?

File: plugins/inputs/mdstat/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+- mdstat
+  - ActivityState (`active` or `inactive`)
+  - Devices (comma separated list of devices that make up the array)
+  - Name (name of the array)
+
+### Example Output:
+
+```
+$ telegraf --config ~/ws/telegraf.conf --input-filter mdstat --test
+* Plugin: kernel_vmstat, Collection 1
```
Comment (ssoroka) [inline plugins/inputs/mdstat/README.md:46]: says vmstat

File: plugins/inputs/mdstat/mdstat.go [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	## Sets file path
+	## If not specified, then default is /proc/mdstat
+	# file_name = "/proc/mdstat"
+`
+
+func (k *MdstatConf) SampleConfig() string {
+	return mdSampleConfig
+}
+
+func evalStatusLine(deviceLine, statusLineStr string) statusLine {
```
Comment (ssoroka) [inline plugins/inputs/mdstat/mdstat.go:80]: by convention, Go typically has an error as a separate return argument

File: plugins/inputs/mdstat/mdstat.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+			err: fmt.Errorf("statusLine empty? %q", statusLineStr)}
+	}
+	sizeStr := sizeFields[0]
+	size, err := strconv.ParseInt(sizeStr, 10, 64)
+	if err != nil {
+		return statusLine{active: 0, total: 0, size: 0,
+			err: fmt.Errorf("unexpected statusLine %q: %w", statusLineStr, err)}
+	}
+
+	if strings.Contains(deviceLine, "raid0") || strings.Contains(deviceLine, "linear") {
```
Comment (ssoroka) [inline plugins/inputs/mdstat/mdstat.go:93]: is it possible this match could be too liberal?

File: plugins/inputs/mdstat/mdstat_test.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+unused devices: <none>
+`
+
+/*
+const mdStatFileEmpty = `
+Personalities :
+unused devices: <none>
+`
+*/
```
Comment (ssoroka) [inline plugins/inputs/mdstat/mdstat_test.go:73]: should there be a test for this? 

if not we can remove the comment.

**Review by ssoroka — CHANGES_REQUESTED** [review]
Looks great. I have some comments below

**Comment by popey** [comment]
Thanks for the new plugin! As a user of Linux mdraid arrays, I'm keen to monitor them!

I ran it on a system here with four disks, one which is failed, and I don't see it detecting the failed disk?

Here's the `/proc/mdstat` output
```
Personalities : [linear] [multipath] [raid0] [raid1] [raid6] [raid5] [raid4] [raid10] 
md0 : active raid5 sdd1[3] sdb1[1] sda1[0]
      5860144128 blocks super 1.2 level 5, 64k chunk, algorithm 2 [4/3] [UUU_]
      bitmap: 8/15 pages [32KB], 65536KB chunk

unused devices: <none>
```

Here's the output:
```
2021-05-25T21:45:53Z I! Starting Telegraf 
2021-05-25T21:45:53Z I! Loaded inputs: mdstat
2021-05-25T21:45:53Z I! Loaded aggregators: 
2021-05-25T21:45:53Z I! Loaded processors: 
2021-05-25T21:45:53Z I! Loaded outputs: file
2021-05-25T21:45:53Z I! Tags enabled: host=shirka
2021-05-25T21:45:53Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"shirka", Flush Interval:10s
mdstat,ActivityState=active,Devices=sda1\,sdb1\,sdd1,Name=md0,host=shirka DisksActive=3i,DisksTotal=4i,BlocksSynced=5860144128i,BlocksSyncedSpeed=0,BlocksSyncedFinishTime=0,DisksFailed=0i,DisksSpare=0i,BlocksTotal=5860144128i,BlocksSyncedPct=0 1621979160000000000

```

File: plugins/inputs/mdstat/mdstat.go (hunk 4 above)
Comment (johnseekins) [reply plugins/inputs/mdstat/mdstat.go:93]: I actually pulled that match line from the prometheus/procps library. If it _is_ too liberal, then it's broken there, too.

**Comment by johnseekins** [comment]
> Thanks for the new plugin! As a user of Linux mdraid arrays, I'm keen to monitor them!
> 
> I ran it on a system here with four disks, one which is failed, and I don't see it detecting the failed disk? Perhaps the disk just dropped but hasn't failed?
> 
> Here's the `/proc/mdstat` output
> 
> ```
> Personalities : [linear] [multipath] [raid0] [raid1] [raid6] [raid5] [raid4] [raid10] 
> md0 : active raid5 sdd1[3] sdb1[1] sda1[0]
>       5860144128 blocks super 1.2 level 5, 64k chunk, algorithm 2 [4/3] [UUU_]
>       bitmap: 8/15 pages [32KB], 65536KB chunk
> 
> unused devices: <none>
> ```
> 
> Here's the output:
> 
> ```
> 2021-05-25T21:45:53Z I! Starting Telegraf 
> 2021-05-25T21:45:53Z I! Loaded inputs: mdstat
> 2021-05-25T21:45:53Z I! Loaded aggregators: 
> 2021-05-25T21:45:53Z I! Loaded processors: 
> 2021-05-25T21:45:53Z I! Loaded outputs: file
> 2021-05-25T21:45:53Z I! Tags enabled: host=shirka
> 2021-05-25T21:45:53Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"shirka", Flush Interval:10s
> mdstat,ActivityState=active,Devices=sda1\,sdb1\,sdd1,Name=md0,host=shirka DisksActive=3i,DisksTotal=4i,BlocksSynced=5860144128i,BlocksSyncedSpeed=0,BlocksSyncedFinishTime=0,DisksFailed=0i,DisksSpare=0i,BlocksTotal=5860144128i,BlocksSyncedPct=0 1621979160000000000
> ```

This should be resolved with the most recent commits.

File: plugins/inputs/mdstat/mdstat.go (hunk 4 above)
Comment (johnseekins) [reply plugins/inputs/mdstat/mdstat.go:92]: And more specifically, that if statement is only looking at the "device" line. That line looks like this:
```
active raid1 sdb1[2] sda1[0]
```
So we should only see one array "type" on the line.

**Comment by johnseekins** [comment]
The same collector was approved and merged into Prometheus' node_exporter: https://github.com/prometheus/procfs/pull/380

**Comment by johnseekins** [comment]
@ssoroka I believe I've addressed all your concerns. Any chance you could take another look at this?

**Review by ssoroka — APPROVED** [review]

**Pull request merged**